src/libmongoc/src/mongoc/mongoc-collection.c

```diff
@@ -825,24 +882,45 @@ mongoc_collection_estimated_document_count (
    }
 
    reply_ptr = reply ? reply : &reply_local;
-   bson_append_utf8 (&cmd, "count", 5, coll->collection, coll->collectionlen);
-
-   ret = _mongoc_client_command_with_opts (coll->client,
-                                           coll->db,
-                                           &cmd,
-                                           MONGOC_CMD_READ,
-                                           opts,
-                                           MONGOC_QUERY_NONE,
-                                           read_prefs,
-                                           coll->read_prefs,
-                                           coll->read_concern,
-                                           coll->write_concern,
-                                           reply_ptr,
-                                           error);
-
-   if (ret) {
-      if (bson_iter_init_find (&iter, reply_ptr, "n")) {
-         count = bson_iter_as_int64 (&iter);
+   if (server_stream->sd->max_wire_version < WIRE_VERSION_4_9) {
+      /* On < 4.9, use actual count command for estimatedDocumentCount */
+      bson_init (&cmd);
+      bson_append_utf8 (&cmd, "count", 5, coll->collection, coll->collectionlen);
+      ret = _mongoc_client_command_with_opts (coll->client,
+                                             coll->db,
+                                             &cmd,
+                                             MONGOC_CMD_READ,
+                                             opts,
+                                             MONGOC_QUERY_NONE,
+                                             read_prefs,
+                                             coll->read_prefs,
+                                             coll->read_concern,
+                                             coll->write_concern,
+                                             reply_ptr,
+                                             error);
+      if (ret) {
+         if (bson_iter_init_find (&iter, reply_ptr, "n")) {
+            count = bson_iter_as_int64 (&iter);
+         }
+      }
+   } else {
```

Comment:
I mentioned this in the other comment and want to leave the final decision to @kevinAlbs and @bazile-clyde, but we may want to consider crafting a `reply` that matches the `count` command reply document when using agg pipeline to determine document count. While technically the details of how we get to a document count are internal and shouldn't be relied on by users, the fact that the reply document changes significantly may present a problem for some users.

I don't think we need to do anything for the error case, but in case we're returning any count we may want to craft a document like `{ "n" : <count>, "ok" : 1 }` to return when using agg pipeline to retrieve document count.

Reply:
It's a great point, and it seems like we should put some thought into the consistency of the returned `reply`. I've created [CDRIVER-3933](https://jira.mongodb.org/browse/CDRIVER-3933) to do this work separately from this PR. The spec only requires us to return an int64 `count` and makes no mention of returning the server response, so we may get rid of `reply` entirely.

Reply:
Sounds good to me, thanks for creating the follow-up ticket!